libraries/core_libs/network/src/tarcap/packets_handlers/vote_packet_handler.cpp

```diff
@@ -30,28 +62,54 @@ void VotePacketHandler::process(const PacketData &packet_data, const std::shared
     const auto vote_hash = vote->getHash();
     LOG(log_dg_) << "Received PBFT vote " << vote_hash;
 
+    // Check period of vote. Request PBFT sync if we have a future one here.
+    if (vote->getPeriod() > pbft_chain_->getPbftChainSize() + kVoteAcceptingPeriods) {
+      LOG(log_er_) << "Skip vote " << vote_hash.abridged() << ". Period is too big " << pbft_chain_->getPbftChainSize()
+                   << " " << vote->getPeriod();
+      // Do not request round sync too often here
+      if (std::chrono::system_clock::now() - round_sync_request_time_ > kSyncRequestInterval) {
+        // request sync from this node
+        sealAndSend(peer->getId(), SubprotocolPacketType::GetPbftSyncPacket,
+                    std::move(dev::RLPStream(1) << vote->getPeriod()));
+        round_sync_request_time_ = std::chrono::system_clock::now();
+      }
+      continue;
+    }
+    // Check this before we are marking vote as seen to have an opportunity to process it after
+    if (!isVoteRoundStepRelevant(vote)) {
+      // Do not request round sync too often here
+      if (std::chrono::system_clock::now() - round_sync_request_time_ > kSyncRequestInterval) {
+        // request sync from this node
+        sealAndSend(
+            peer->getId(), GetVotesSyncPacket,
+            std::move(dev::RLPStream(3) << pbft_chain_->getPbftChainSize() + 1 << pbft_mgr_->getPbftRound() << 0));
+        round_sync_request_time_ = std::chrono::system_clock::now();
+      }
+      continue;
+    }
+
     // Synchronization point in case multiple threads are processing the same vote at the same time
     if (!seen_votes_.insert(vote_hash)) {
       LOG(log_dg_) << "Received vote " << vote_hash << " (from " << packet_data.from_node_id_.abridged()
                    << ") already seen.";
       continue;
     }
-    peer->markVoteAsKnown(vote_hash);
 
     // Standard vote
     if (vote->getPeriod() >= current_pbft_period) {
       if (vote_mgr_->voteInVerifiedMap(vote)) {
         LOG(log_dg_) << "Vote " << vote_hash.abridged() << " already inserted in verified queue";
-      }
-
-      if (auto vote_is_valid = validateStandardVote(vote); vote_is_valid.first == false) {
-        LOG(log_wr_) << "Vote " << vote_hash.abridged() << " validation failed. Err: " << vote_is_valid.second;
-        continue;
-      }
-
-      if (!vote_mgr_->addVerifiedVote(vote)) {
-        LOG(log_dg_) << "Vote " << vote_hash << " already inserted in verified queue(race condition)";
-        continue;
+        // We don't need to validate and process it, but should add to votes for gossiping
```

Comment:
we could just do:

```
if (!vote_mgr_->voteInVerifiedMap(vote)) {
  // validate....

}
```

without any else ?

Reply:
But do we need to call `validateStandardVote` and `addVerifiedVote` if this vote is already in verified map?

Reply:
true, we dont - there is !

```
if (!vote_mgr_->voteInVerifiedMap(vote)) {
  // validate....

}
```

Reply:
I mean it does the same thing, I just find the if else not that useful as it can be done with single if but choose what you like more

Reply:
Yep. I missed that  added `!` in your comment, so I missed all the logic behind it. And this log inside shouldn't be very useful, so I will change this